File: frontend/src/api/glid.ts
// glid.ts — frontend GLID encoding/decoding.
// Matches backend internal/glid: base32hex (RFC 4648), lowercase, no padding.
// 16 raw bytes ↔ 26-char string, lexicographically sortable by UUIDv7 time.

const ALPHABET = "0123456789abcdefghijklmnopqrstuv";
const DECODE_MAP = new Uint8Array(128);
for (let i = 0; i < ALPHABET.length; i++) {
  DECODE_MAP[ALPHABET.codePointAt(i)!] = i;
  DECODE_MAP[ALPHABET.toUpperCase().codePointAt(i)!] = i;
}

const SIZE = 16;
const ENCODED_LEN = 26;
const ZERO = new Uint8Array(SIZE);

/** Encode 16 raw bytes to a 26-char base32hex string (lowercase, no padding). */
export function encode(bytes: Uint8Array): string {
  if (bytes.length < SIZE) return "";
  let allZero = true;
  for (let i = 0; i < SIZE; i++) {
    if (bytes[i] !== 0) { allZero = false; break; }
  }
  if (allZero) return "";

  // Base32hex: 5 bytes → 8 chars. 16 bytes = 3×5 + 1 remainder byte.
  const chars: string[] = [];
  let i = 0;
  // Process 3 full 5-byte groups (15 bytes → 24 chars)
  for (let g = 0; g < 3; g++) {
    const b0 = bytes[i++]!, b1 = bytes[i++]!, b2 = bytes[i++]!, b3 = bytes[i++]!, b4 = bytes[i++]!;
    chars.push(
      ALPHABET[(b0 >> 3)]!,
      ALPHABET[((b0 & 0x07) << 2) | (b1 >> 6)]!,
      ALPHABET[(b1 >> 1) & 0x1F]!,
      ALPHABET[((b1 & 0x01) << 4) | (b2 >> 4)]!,
      ALPHABET[((b2 & 0x0F) << 1) | (b3 >> 7)]!,
      ALPHABET[(b3 >> 2) & 0x1F]!,
      ALPHABET[((b3 & 0x03) << 3) | (b4 >> 5)]!,
      ALPHABET[b4 & 0x1F]!,
    );
  }
  // Remaining 1 byte → 2 chars (no padding)
  const b0 = bytes[i]!;
  chars.push(
    ALPHABET[(b0 >> 3)]!,
    ALPHABET[((b0 & 0x07) << 2)]!,
  );

  return chars.join("");
}

/** Decode a 26-char base32hex string to 16 raw bytes. Returns zero bytes for empty/invalid input. */
export function decode(s: string): Uint8Array<ArrayBuffer> {
  if (s?.length !== ENCODED_LEN) return new Uint8Array(ZERO);

  const out = new Uint8Array(SIZE);
  let oi = 0;
  let si = 0;

  // 3 full 5-byte groups from 24 chars
  for (let g = 0; g < 3; g++) {
    const c0 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c1 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c2 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c3 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c4 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c5 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c6 = DECODE_MAP[s.codePointAt(si++)!]!;
    const c7 = DECODE_MAP[s.codePointAt(si++)!]!;
    out[oi++] = (c0 << 3) | (c1 >> 2);
    out[oi++] = ((c1 & 0x03) << 6) | (c2 << 1) | (c3 >> 4);
    out[oi++] = ((c3 & 0x0F) << 4) | (c4 >> 1);
    out[oi++] = ((c4 & 0x01) << 7) | (c5 << 2) | (c6 >> 3);
    out[oi++] = ((c6 & 0x07) << 5) | c7;
  }
  // Remaining 2 chars → 1 byte
  const c0 = DECODE_MAP[s.codePointAt(si++)!]!;
  const c1 = DECODE_MAP[s.codePointAt(si)!]!;
  out[oi] = (c0 << 3) | (c1 >> 2);

  return out;
}

/** Check if bytes represent the zero GLID. */
export function isZero(bytes: Uint8Array): boolean {
  if (bytes.length < SIZE) return true;
  for (let i = 0; i < SIZE; i++) {
    if (bytes[i] !== 0) return false;
  }
  return true;
}

File: backend/internal/glid/glid.go
// Package glid provides a unified identifier type for all GastroLog entities.
// GLID wraps uuid.UUID (v7) and adds base32hex string encoding (26 chars,
// lexicographically sortable by creation time). On the proto wire format,
// GLIDs are raw bytes (16 bytes, no encoding overhead).
package glid

import (
	"encoding/base32"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/google/uuid"
)

// Size is the byte length of a GLID (same as UUID: 16 bytes).
const Size = 16

// encoding is base32hex (RFC 4648) lowercase without padding.
// Alphabet 0-9a-v preserves lexicographic sort order.
var encoding = base32.HexEncoding.WithPadding(base32.NoPadding)

// GLID uniquely identifies an entity. It wraps uuid.UUID (v7) with
// base32hex string encoding. Sortable by creation time.
type GLID uuid.UUID

// Nil is the zero GLID.
var Nil GLID

// New creates a new GLID from a fresh UUIDv7.
func New() GLID {
	return GLID(uuid.Must(uuid.NewV7()))
}

// FromUUID converts a uuid.UUID to a GLID.
func FromUUID(u uuid.UUID) GLID {
	return GLID(u)
}

// UUID returns the underlying uuid.UUID.
func (g GLID) UUID() uuid.UUID {
	return uuid.UUID(g)
}

// Parse parses a 26-character base32hex string into a GLID.
func Parse(s string) (GLID, error) {
	if len(s) == 0 {
		return Nil, nil
	}
	if len(s) != 26 {
		return Nil, fmt.Errorf("invalid GLID length: %d (want 26)", len(s))
	}
	decoded, err := encoding.DecodeString(strings.ToUpper(s))
	if err != nil {
		return Nil, fmt.Errorf("invalid GLID: %w", err)
	}
	var g GLID
	copy(g[:], decoded)
	return g, nil
}

// ParseUUID parses a UUID string or base32hex string into a GLID.
// Accepts both formats for backward compatibility.
func ParseUUID(s string) (GLID, error) {
	return ParseAny(s)
}

// FromBytes creates a GLID from a 16-byte slice (proto wire format).
// Returns Nil for nil/empty/short input.
func FromBytes(b []byte) GLID {
	if len(b) < Size {
		return Nil
	}
	var g GLID
	copy(g[:], b)
	return g
}

// Bytes returns the raw 16-byte representation (proto wire format).
func (g GLID) Bytes() []byte {
	return g[:]
}

// String returns the 26-character lowercase base32hex representation.
func (g GLID) String() string {
	if g == Nil {
		return ""
	}
	return strings.ToLower(encoding.EncodeToString(g[:]))
}

// IsZero reports whether the GLID is the zero value.
func (g GLID) IsZero() bool {
	return g == Nil
}

// Time delegates to uuid.UUID.Time(), which returns the UUIDv7 timestamp
// with the full precision that the uuid package provides.
func (g GLID) Time() uuid.Time {
	return uuid.UUID(g).Time()
}

// MarshalJSON encodes the GLID as a JSON string in base32hex format.
func (g GLID) MarshalJSON() ([]byte, error) {
	return json.Marshal(g.String())
}

// UnmarshalJSON decodes a JSON string in base32hex format.
func (g *GLID) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	parsed, err := Parse(s)
	if err != nil {
		return err
	}
	*g = parsed
	return nil
}

// MarshalText encodes the GLID as base32hex text.
func (g GLID) MarshalText() ([]byte, error) {
	return []byte(g.String()), nil
}

// UnmarshalText decodes base32hex text.
func (g *GLID) UnmarshalText(data []byte) error {
	parsed, err := Parse(string(data))
	if err != nil {
		return err
	}
	*g = parsed
	return nil
}

// Compare returns -1, 0, or 1 for ordering. UUIDv7 GLIDs are naturally
// ordered by creation time.
func (g GLID) Compare(other GLID) int {
	for i := range Size {
		if g[i] < other[i] {
			return -1
		}
		if g[i] > other[i] {
			return 1
		}
	}
	return 0
}

// ParseOptional parses a base32hex string, returning a pointer.
// Returns nil for empty strings.
func ParseOptional(s string) (*GLID, error) {
	if s == "" {
		return nil, nil
	}
	g, err := Parse(s)
	if err != nil {
		return nil, err
	}
	return &g, nil
}

// OptionalString returns the base32hex string or empty for nil.
func OptionalString(g *GLID) string {
	if g == nil {
		return ""
	}
	return g.String()
}

// MustParse parses a base32hex or UUID string or panics. For tests and constants.
func MustParse(s string) GLID {
	g, err := ParseAny(s)
	if err != nil {
		panic(err)
	}
	return g
}

// ParseAny tries base32hex (26 chars) first, then UUID format (36 chars).
// Accepts both formats for backward compatibility during migration.
func ParseAny(s string) (GLID, error) {
	if s == "" {
		return Nil, nil
	}
	if len(s) == 26 {
		return Parse(s)
	}
	u, err := uuid.Parse(s)
	if err != nil {
		return Nil, err
	}
	return GLID(u), nil
}

// --- Proto boundary helpers ---
// These simplify the conversion between glid.GLID and proto []byte fields.

// ToProto converts a GLID to the proto wire format ([]byte).
// Alias for Bytes() — exists for readability at proto boundaries.
func (g GLID) ToProto() []byte { return g[:] }

// OptionalToProto converts a *GLID to proto bytes. Returns nil for nil pointer.
func OptionalToProto(g *GLID) []byte {
	if g == nil {
		return nil
	}
	return g[:]
}

// SliceToProto converts []GLID to [][]byte for repeated proto fields.
func SliceToProto(ids []GLID) [][]byte {
	if len(ids) == 0 {
		return nil
	}
	out := make([][]byte, len(ids))
	for i, id := range ids {
		out[i] = id[:]
	}
	return out
}

// SliceFromProto converts [][]byte from repeated proto fields to []GLID.
func SliceFromProto(bs [][]byte) []GLID {
	if len(bs) == 0 {
		return nil
	}
	out := make([]GLID, len(bs))
	for i, b := range bs {
		out[i] = FromBytes(b)
	}
	return out
}

// OptionalFromProto converts proto bytes to *GLID. Returns nil for nil/empty input.
func OptionalFromProto(b []byte) *GLID {
	if len(b) < Size {
		return nil
	}
	g := FromBytes(b)
	return &g
}

var (
	ErrInvalidLength = errors.New("invalid GLID length")
	ErrInvalidFormat = errors.New("invalid GLID format")
)
